src/transactions/test/InvokeHostFunctionTests.cpp

```diff
@@ -3533,6 +3773,297 @@ TEST_CASE("persistent entry archival", "[tx][soroban][archival]")
+TEST_CASE("autorestore contract instance", "[tx][soroban][archival]")
+{
+    auto cfg = getTestConfig();
+    SorobanTest test(cfg, true, [](SorobanNetworkConfig& cfg) {
+        cfg.stateArchivalSettings().minPersistentTTL =
+            MinimumSorobanNetworkConfig::MINIMUM_PERSISTENT_ENTRY_LIFETIME;
+
+        // Never snapshot bucket list so we have stable rent fees
+        cfg.mStateArchivalSettings.liveSorobanStateSizeWindowSamplePeriod =
+            10'000;
+    });
+
+    ContractStorageTestClient client(test);
+
+    client.put("key", ContractDataDurability::PERSISTENT, 123);
+
+    auto expirationLedger =
+        test.getLCLSeq() +
+        MinimumSorobanNetworkConfig::MINIMUM_PERSISTENT_ENTRY_LIFETIME;
+
+    // Close ledgers until ContractData entry, contract code, and instance are
+    // all expired
+    for (uint32_t ledgerSeq = test.getLCLSeq() + 1;
+         ledgerSeq <= expirationLedger; ++ledgerSeq)
+    {
+        closeLedgerOn(test.getApp(), ledgerSeq, 2, 1, 2016);
+    }
+
+    auto lk = client.getContract().getDataKey(
+        makeSymbolSCVal("key"), ContractDataDurability::PERSISTENT);
+
+    // We need to restore instance, wasm, and data entry
+    auto const refundableRestoreCost = 60'146;
+    auto keysToRestore = client.getContract().getKeys();
+    keysToRestore.push_back(lk);
+
+    SECTION("manual restore")
+    {
+        test.invokeRestoreOp(keysToRestore, refundableRestoreCost);
+
+        // Contract and entry should be restored
+        client.has("key", ContractDataDurability::PERSISTENT, 123);
+    }
+
+    SECTION("autorestore fees")
+    {
+        // Now submit an invocation with all keys in the write footprint so
+        // everything is restored
+        auto spec = client.defaultSpecWithoutFootprint()
+                        .setReadWriteFootprint(keysToRestore)
+                        .setArchivedIndexes({0, 1, 2})
+                        .setWriteBytes(10000)
+                        .setRefundableResourceFee(70'000);
+
+        auto invocation = client.getContract().prepareInvocation(
+            "put_persistent", {makeSymbolSCVal("key"), makeU64SCVal(123)}, spec,
+            /*addContractKeys=*/false);
+        auto autorestoreTx =
+            invocation.withExactNonRefundableResourceFee().createTx();
+
+        auto initialBalance = test.getAccountBalance();
+        auto txm = TransactionMetaFrame(test.getLedgerVersion(),
+                                        test.getApp().getConfig());
+        REQUIRE(test.invokeTx(autorestoreTx, &txm).result.code() ==
+                TransactionResultCode::txSUCCESS);
+
+        // Check that the refundable fee charged matches what we expect for
+        // restoration
+        auto const eventSize = 4;
+        test.checkRefundableFee(initialBalance, autorestoreTx, txm,
+                                refundableRestoreCost, eventSize);
+
+        // Entry should be live again
+        client.has("key", ContractDataDurability::PERSISTENT, 123);
```

Comment:
This is fascinating because the `has` actually expects `optional<bool>` as the third argument, and your ints get implicitly casted to it (yay, C++!). But I believe you meant `get` here and in a few similar places around.
Also, I think the check that this doesn't work before restoration would be useful, as currently it's not obvious if the test spec doesn't do anything about autorestore.

Reply:
Ahh, good catch. I've changed these back to `get` and added a check before the restore calls.